epsagon/modules/gunicorn.py

```
@@ -0,0 +1,191 @@
+        if hasattr(request, 'headers'):
+            headers = dict(request.headers)
+            self.resource['name'] = (
+                headers.get('HOST', 'localhost').split(':')[0]
+            )
+            add_data_if_needed(
+                self.resource['metadata'],
+                'Request Headers',
+                headers
+            )
+
+        self.resource['operation'] = request.method
+        self.resource['metadata'].update({'Path': request.path})
+        self.resource['metadata'].update({'Query': request.query})
+
+    def update_response(self, response):
+        """
+        Adds response data to event.
+        :param response: WSGI Response
+        :return: None
+        """
+        add_data_if_needed(
+            self.resource['metadata'],
+            'Response Headers',
+            dict(response.headers)
+        )
+
+        if response.status_code >= 500:
+            self.set_error()
+
+
+class PreRequestWrapped(PreRequest):
+    """
+    Represents Gunicorn Middleware's PreRequest Hook
+    """
+    name = 'pre_request'
+    section = 'Server Hooks'
+    validator = validate_callable(2)
+    type = callable
+
+    # pylint: disable=no-self-argument
+    def pre_request(worker, req):
+        """
+        Runs before process of response.
+        """
+        if 'SyncWorker' not in type(worker).__name__:
+            return
+
+        trace = epsagon.trace.trace_factory.get_or_create_trace()
+        trace.prepare()
+
+        if ignore_request('', req.path.lower()):
+            return
+
+        # Create a Gunicorn runner with current request.
+        try:
+            runner = GunicornRunner(
+                time.time(),
+                req
+            )
+            trace.set_runner(runner)
+
+            # Collect metadata in case this is a container.
+            collect_container_metadata(runner.resource['metadata'])
+        # pylint: disable=W0703
+        except Exception as exception:
```

Looks like an Epsagon error will be created for instrumentation failure as well as failure in the original gunicorn code. Can we separate those? (As one is something the customer is interested in, and the other is internal and we want to know about it)

---

That's not true. These code only runs as part of a server hook, meaning it has zero effect on the customers code.

But thank you for mentioning that it made me realize I'm not catching the customer's code exception due to the design of those server hooks.

There's a bug in the gunicorn library that exception are not being propagated to the server hooks, I'm going to have to find a different method to hook.